scipy/optimize/jonkervolgenant/jonkervolgenant.cpp

```diff
@@ -0,0 +1,194 @@
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+This code implements the Jonker-Volgenant algorithm for the rectangular
+assignment problem:
+
+    R. Jonker and A. Volgenant. A Shortest Augmenting Path Algorithm for
+    Dense and Sparse Linear Assignment Problems. *Computing*, 38:325-340
+    December 1987.
+
+The code is based on the pseudocode described in pages
+1685-1686 of:
+
+    DF Crouse. On implementing 2D rectangular assignment algorithms.
+    IEEE Transactions on Aerospace and Electronic Systems
+    52(4):1679-1696, August 2016
+    doi: 10.1109/TAES.2016.140952
+
+Author: PM Larsen
+*/
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+static int
+augmenting_path(int nc, std::vector<double>& cost, std::vector<double>& u,
+                std::vector<double>& v, std::vector<int>& path,
+                std::vector<int>& row4col,
+                std::vector<double>& shortestPathCosts, int i,
+                std::vector<bool>& SR, std::vector<bool>& SC, double* p_minVal)
+{
+    double minVal = 0;
+
+    int num_remaining = nc;
+    std::vector<int> remaining(nc);
+    // std::iota(remaining.begin(), remaining.end(), 0); //needs C++11
+    for (int it = 0; it < nc; it++) {
+        remaining[it] = it;
+    }
+
+    std::fill(SR.begin(), SR.end(), false);
+    std::fill(SC.begin(), SC.end(), false);
+    std::fill(shortestPathCosts.begin(), shortestPathCosts.end(), INFINITY);
+
+    // find shortest augmenting path
+    int sink = -1;
+    while (sink == -1) {
+
+        int index = -1;
+        double lowest = INFINITY;
+
+        for (int it = 0; it < num_remaining; it++) {
+            int j = remaining[it];
+
+            double r = minVal + cost[i * nc + j] - u[i] - v[j];
+            if (r < shortestPathCosts[j]) {
+                path[j] = i;
+                shortestPathCosts[j] = r;
+            }
+
+            if (shortestPathCosts[j] < lowest ||
+                (shortestPathCosts[j] == lowest && row4col[j] == -1)) {
+                lowest = shortestPathCosts[j];
+                index = it;
+            }
+        }
+
+        minVal = lowest;
+        int j = remaining[index];
+        if (minVal == INFINITY) { // infeasible cost matrix
+            return -1;
+        }
+
+        if (row4col[j] == -1) {
+            sink = j;
+        } else {
+            i = row4col[j];
+        }
+
+        SR[i] = true;
```

Comment:
A priori this looks like a discrepancy from the paper this refers to, in which `SR[i]` is set to `true` at the beginning of the loop, so that if `i` changes in the first iteration, we won't be toggling `SR[curRow]`. Didn't quite get through all of it yet, but I imagine that since this line has been explicitly moved, this can simply never happen in practice (i.e. `row4col[j]` is always -1 in the first iteration)? Figured I ought to leave a comment regardless. Relatedly, would it make sense to only set `SR[i]` when `i` actually changes (or does that perhaps also boil down to the same thing)?

Reply:
I believe I understand the logic: We can indeed end up in a situation where `SR[curRow]` ends up with a different value than what it would have if we set it at the beginning of the loop, but this isn't an issue since `curRow` gets special treatment when updating the dual below, so that the value of `SR[curRow]` ends up not mattering.

Reply:
Well spotted.  This was an unintentional deviation from the pseudocode.  I have fixed it now.